src/Npgsql/ClusterAwareDataSource.cs

```diff
@@ -0,0 +1,489 @@
+    /// <summary>
+    /// list of yb_server hosts
+    /// </summary>
+    protected List<string>? _hosts = null;
+    // volatile int _roundRobinIndex = -1;
+    DateTime _lastServerFetchTime = new DateTime(0);
+    readonly double REFRESH_LIST_SECONDS;
+    readonly int MAX_REFRESH_INTERVAL = 600;
+    /// <summary>
+    /// List of unreachable hosts
+    /// </summary>
+    protected List<int> unreachableHostsIndices = new List<int>();
+    /// <summary>
+    /// Stores a boolean value for which IP Address is to be used - Public or Private
+    /// True = Private IPs
+    /// False = Public IPs
+    /// </summary>
+    protected bool? UseHostColumn = null;
+    static int index = 0;
+
+    /// <summary>
+    /// Stores a map of pool index to number of connections made to the pool
+    /// Key = Pool Index
+    /// Value = Number of Connections to that pool
+    /// </summary>
+    protected static Dictionary<int, int> poolToNumConnMap = new Dictionary<int, int>();
+    /// <summary>
+    /// Connection settings
+    /// </summary>
+    protected NpgsqlConnectionStringBuilder settings;
+    internal NpgsqlDataSourceConfiguration dataSourceConfig;
+
+    /// <summary>
+    /// List of initial hosts for control connections
+    /// </summary>
+    protected List<string>? initialHosts;
+
+    /// <summary>
+    /// Lock object
+    /// </summary>
+    protected static readonly object lockObject = new object();
+
+    internal ClusterAwareDataSource(NpgsqlConnectionStringBuilder settings, NpgsqlDataSourceConfiguration dataSourceConfig, bool useClusterAwareDataSource)
+        : base(settings, dataSourceConfig)
+    {
+        this.settings = settings;
+        this.dataSourceConfig = dataSourceConfig;
+        REFRESH_LIST_SECONDS = settings.YBServersRefreshInterval > 0 && settings.YBServersRefreshInterval < MAX_REFRESH_INTERVAL
+            ? this.settings.YBServersRefreshInterval
+            : 300;
+        initialHosts = Settings.Host?.Split(',').ToList();
+        Debug.Assert(initialHosts != null, nameof(initialHosts) + " != null");
+        if(useClusterAwareDataSource)
+        {
+            instance = this;
+            foreach (var host in initialHosts.ToList())
+            {
+                try
+                {
+                    var controlSettings = settings;
+                    controlSettings.Host = host.ToString();
+                    NpgsqlDataSource control = new UnpooledDataSource(controlSettings, dataSourceConfig);
+                    NpgsqlConnection controlConnection = NpgsqlConnection.FromDataSource(control);
+                    controlConnection.Open();
+                    CreatePool(controlConnection);
+                    controlConnection.Close();
+                    break;
+                }
+                catch (Exception)
+                {
+                    initialHosts.Remove(host);
+                    if (initialHosts.Count == 0)
+                        throw;
```

Comment:
Should throw an exception with proper message.

Reply:
This rethrows the caught exception